### PR TITLE
general-concepts/licenses: Add a section on upstream license problems

### PR DESCRIPTION
**Pull request by mgorny**
Signed-off-by: Michał Górny <mgorny@gentoo.org>

### CODE REVIEW

File: general-concepts/licenses/text.xml [hunk 1]
```diff
@@ -81,14 +81,105 @@ of applicable licenses but the exact application of licenses and their
...
+      Including third party code without appropriate copyright notices.
+      Practically all licenses (with notable exception of public
+      domain-alike) require attribution, and some require copying
+      original copyright notices verbatim.
+    </p>
+
+    <p>
+      This issue can be fixed by readding the original copyright
+      notices. Please note that the license of the project may need
+      to be reevaluated afterwards.
```
Comment (ulm) [inline general-concepts/licenses/text.xml]: Copyright notices should only be added upstream, but never by the Gentoo maintainer (except he's upstream himself).

File: general-concepts/licenses/text.xml [hunk 2]
```diff
@@ -81,14 +81,105 @@ of applicable licenses but the exact application of licenses and their
...
+    </p>
+  </li>
+
+  <li>
+    <p>
+      Combining incompatible licenses. When you are combining multiple
+      files using different licenses into a single executable, those
+      licenses need to be compatible. For example, it is not possible
+      to combine proprietary code with copyleft licenses (e.g. GPL).
+      But it also incorrect to combine GPL-2 and GPL-3 code.
```
Comment (ulm) [inline general-concepts/licenses/text.xml]: This sentence no verb.

File: general-concepts/licenses/text.xml [hunk 3]
```diff
@@ -81,14 +81,105 @@ of applicable licenses but the exact application of licenses and their
...
+    </p>
+
+    <p>
+      In some cases, upstream may be able to resolve the issue by
+      relicensing parts of the code to a compatible license. However,
+      this requires permission from copyright holders. If this is
+      not feasible, the incompatible code can be either removed or
+      replaced with an alternative implementation. If the original code
+      is independently rewritten, a clear room design is recommended.
```
Comment (ulm) [inline general-concepts/licenses/text.xml]: "clean room" (not "clear room").

File: general-concepts/licenses/text.xml [hunk 4]
```diff
@@ -81,14 +81,105 @@ of applicable licenses but the exact application of licenses and their
...
+    </p>
+  </li>
+
+  <li>
+    <p>
+      Dynamically linking incompatible executables. In addition to
+      restrictions made on combining code into a single executable, some
+      licenses also apply restriction on dynamic linking between
+      executables and shared libraries. For example, normally you can't
```
Comment (ulm) [inline general-concepts/licenses/text.xml]: IIUC it is still an open question if licenses can restrict dynamic linking. See for example Lawrence Rosen's book (https://www.rosenlaw.com/oslbook.htm, chapter 6, page 115 ff), or his article in the Linux Journal: https://www.linuxjournal.com/article/6366

So maybe use a more careful wording, e.g. "arguably, some licenses apply restrictions ...".

**Comment by mgorny** [comment]
Ping.

**Comment by gktrk** [comment]
I'm okay with the parts that help developers detect potential license issues. I'm not comfortable listing potential solutions to them in devmanual. It is not our job to tell upstream how to fix their license problems.

I don't mind if a developer offers upstream help. I just object that they should feel empowered to do it based on the few paragraphs that they read here.

So I propose merging only the parts on detecting license problems.

**Pull request closed without merging**